lib/pbench/agent/tool_data_sink.py

```diff
@@ -932,308 +1402,440 @@ def put_document(self, data_ctx, hostname):
         Bottle abort() method for error handling.
 
         """
-        with self._lock:
-            if self.state not in self._data_states:
-                # FIXME: Don't we have a race condition if the tool meisters
-                # process their messages first?  Seems like we need to send to
-                # the Tool Data Sink first, and then send to all the tool
-                # meisters.
-                abort(400, f"Can't accept PUT requests in state '{self.state}'")
-            if self.data_ctx != data_ctx:
-                # Tool Data Sink and this Tool Meister are out of sink as to
-                # what data is expected.
-                abort(400, f"Unexpected data context, '{data_ctx}'")
-            # Fetch the Tool Meister tracking record for this host and verify
-            # it is in the expected waiting state.
-            tm_tracker = self._tm_tracking[hostname]
-            if tm_tracker["posted"] != "waiting":
-                self.logger.error(
-                    "INTERNAL ERROR: expected Tool Meister for host, %s, in"
-                    " `waiting` state, found in `%s` state",
-                    hostname,
-                    tm_tracker["posted"],
-                )
-                abort(400, "No data expected from a Tool Meister")
-
-        try:
-            content_length = int(request["CONTENT_LENGTH"])
-        except ValueError:
-            abort(400, "Invalid content-length header, not an integer")
-        except Exception:
-            abort(400, "Missing required content-length header")
-        else:
-            if content_length > _MAX_TOOL_DATA_SIZE:
-                abort(
-                    400,
-                    "Content object too large, keep it at 1 GB"
-                    f" ({content_length:d}) and  under",
-                )
-            remaining_bytes = content_length
-
         try:
-            exp_md5 = request["HTTP_MD5SUM"]
-        except Exception:
-            self.logger.exception(request.keys())
-            abort(400, "Missing required md5sum header")
+            with self._lock:
+                if self.action not in self._data_actions:
+                    abort(400, f"Can't accept PUT requests in action '{self.action}'")
+                if self.data_ctx != data_ctx:
+                    # Tool Data Sink and this Tool Meister are out of sink as to
```

Comment:
They are "out of sync" not "out of sink"!

(The perhaps one good thing about re-reviewing files on github is that I may tend to re-read more stuff changed in earlier commits I'd already seen...)

Reply:
Still sinking... going down fast ...